pkg/ippoolmanager/ippool_informer.go

```diff
@@ -157,22 +192,233 @@ func (im *ipPoolManager) updateSpiderIPPool(ctx context.Context, oldSpiderIPPool
+}
+
+func (c *controller) checkScaleIPPool(ctx context.Context, pool *spiderpoolv1.SpiderIPPool) error {
+	poolLabels := pool.GetLabels()
+	subnetName, ok := poolLabels[constant.LabelIPPoolOwnerSpiderSubnet]
+	if !ok {
+		return fmt.Errorf("%w: there's no owner SpiderSubnet for IPPool '%s'", constant.ErrWrongInput, pool.Name)
+	}
+
+	if pool.Status.AutoDesiredIPCount == nil {
+		informerLogger.Sugar().Debugf("maybe IPPool '%s' is just created for a while, wait for updating status DesiredIPCount", pool.Name)
+		return nil
+	}
+
+	totalIPs, err := spiderpoolip.ParseIPRanges(*pool.Spec.IPVersion, pool.Spec.IPs)
+	if nil != err {
+		return fmt.Errorf("%w: failed to parse IP ranges '%v', error: %v", constant.ErrWrongInput, pool.Spec.IPs, err)
+	}
+
+	desiredIPNum := int(*pool.Status.AutoDesiredIPCount)
+	totalIPCount := len(totalIPs)
+
+	if desiredIPNum == totalIPCount {
+		// no need to scale
+		return nil
+	}
+
+	pool = pool.DeepCopy()
+
+	if desiredIPNum > totalIPCount {
+		// expand
+		ipsFromSubnet, err := c.poolMgr.subnetManager.GenerateIPsFromSubnet(ctx, subnetName, desiredIPNum-totalIPCount)
+		if nil != err {
+			return fmt.Errorf("failed to generate IPs from subnet '%s', error: %v", subnetName, err)
+		}
+
+		informerLogger.Sugar().Infof("try to scale IPPool '%s' IP number from '%d' to '%d' with generated IPs '%v'", pool.Name, totalIPCount, desiredIPNum, ipsFromSubnet)
+		err = c.poolMgr.ScaleIPPoolWithIPs(ctx, pool, ipsFromSubnet, ippoolmanagertypes.Expansion, desiredIPNum)
+		if nil != err {
+			return fmt.Errorf("failed to expand IPPool '%s' with IPs '%v', error: %v", pool.Name, ipsFromSubnet, err)
+		}
+	} else {
+		// shrink: free IP number >= return IP Num
+		if totalIPCount-len(pool.Status.AllocatedIPs) >= totalIPCount-desiredIPNum {
+			var allocatedIPRanges []string
+			for tmpIP := range pool.Status.AllocatedIPs {
+				allocatedIPRanges = append(allocatedIPRanges, tmpIP)
+			}
+
+			allocatedIPs, err := spiderpoolip.ParseIPRanges(*pool.Spec.IPVersion, allocatedIPRanges)
+			if nil != err {
+				return fmt.Errorf("%w: failed to parse IP ranges '%v', error: %v", constant.ErrWrongInput, allocatedIPRanges, err)
+			}
+			freeIPs := spiderpoolip.IPsDiffSet(totalIPs, allocatedIPs)
+			// sort freeIPs
+			sort.Slice(freeIPs, func(i, j int) bool {
+				return bytes.Compare(freeIPs[i].To16(), freeIPs[j].To16()) < 0
+			})
+			discardedIPs := freeIPs[:totalIPCount-desiredIPNum]
+			discardedIPRanges, err := spiderpoolip.ConvertIPsToIPRanges(*pool.Spec.IPVersion, discardedIPs)
+			if nil != err {
+				return fmt.Errorf("%w: failed to convert IPs '%v' to IP ranges, error: %v", constant.ErrWrongInput, discardedIPs, err)
+			}
+
+			informerLogger.Sugar().Infof("try to scale IPPool '%s' IP number from '%d' to '%d' with discarded IPs '%v'", pool.Name, totalIPCount, desiredIPNum, discardedIPs)
+			err = c.poolMgr.ScaleIPPoolWithIPs(logutils.IntoContext(ctx, informerLogger), pool, discardedIPRanges, ippoolmanagertypes.Shrinkage, desiredIPNum)
```

Comment:
// comment: the ippool webhook will automatically assign the scaled IP from spidersubnet